### ISSUE TITLE
Confused by the bottle_neck name in retrain tutorial and the inception-v3 model definition

### ISSUE THREAD

**Issue by universewill** [tensorflow/tensorflow#6084]
Inception-v3 model definition link: https://github.com/tensorflow/models/blob/master/inception/inception/slim/inception_model.py

The 'retrain.py' tutorial link: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/image_retraining/retrain.py

In the retrain.py, I find the bottle_neck name of inception3 model is 'pool_3/_reshape:0' (line 95), but in the inception model definition file i can not find definition of 'pool3'.

Also I load the file 'http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz', and get the bottle_neck tensor by using name 'pool_3:0'. And it is a 4 dimension tensor instead of 2.

What is the real name of bottle_neck tensor? Why can not find the bottle_name in model definition python file?

**Comment by shlens** [comment]
The inception model described in this file:
https://github.com/tensorflow/models/blob/master/inception/inception/slim/inception_model.py
corresponds to this tar-ball
http://download.tensorflow.org/models/image/imagenet/inception-v3-2016-03-01.tar.gz

The inception model employed by retrain.py here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/image_retraining/retrain.py
is a *different and older* version of the model. That model is
http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz

**Comment by universewill** [comment]
Thank a lot. Besides, how to find the corresponding model definitions of tar-ball and the model definition of the retrain.py?

**Comment by shlens** [comment]
@petewarden Pete, would you have know the corresponding model definition for the tar-ball version of Inception-v3 used by retrain.py?

**Comment by universewill** [comment]
I download the inception-v3-2016-03-01.tar.gz and find the model file is 'model.ckpt-157585'.
However, the model file in inception-2015-12-05.tgz is a '.pb' file named as 'classify_image_graph_def.pb'. 

Is the 'model.ckpt' model file is created by slim while 'classify_image_graph_def.pb' not?

**Comment by petewarden** [comment]
The GraphDef file you need is currently not available from slim unfortunately. We should be able to do something similar to export_for_serving, but with `f.write(inference_graph.as_graph_def().SerializeToString())` to write out the correct file, but nobody's added that yet.

**Comment by AihamTaleb** [comment]
Hi, thanks for your support guys. Is the file mentioned by @petewarden available now?
Would be possible to tell please where I can find the mapping between the old layer/node/tensor names and the new ones? e.g. what's the corresponding tensor for "pool_3/_reshape:0"?

**Comment by petewarden** [comment]
The file is now available at https://storage.googleapis.com/download.tensorflow.org/models/inception_v3_2016_08_28_frozen.pb.tar.gz

To figure out the node names to use, try the summarize_graph tool:
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/graph_transforms#inspecting-graphs

This will tell you the input and output names, and the `--print_structure` flag will give you a full listing of all the nodes. If you look for the inputs to the final output node, you should be able to figure out the right layer name. If not, paste the output here and I can take a look.

**Comment by AihamTaleb** [comment]
@petewarden Thanks! I could manage to make it work.

**Comment by pronot** [comment]
@AihamTaleb I have the same issue, but still couldn't figure out how to solve it. Could you share your solution? e.g. the corresponding tensor for "pool_3/_reshape:0"?

**Comment by AihamTaleb** [comment]
@pronot Well, I didn't need the name of the corresponding tensor. For the model in this file: https://github.com/tensorflow/models/blob/master/inception/inception/slim/inception_model.py 

I did the following:

- I constructed inception_v3 model only until the avg_pool layer. (Line 320). And then returned the net. This operation is the one that gives you the 2048 vector for each image. Because all consequent operations are related to the actual classification.
- Then, you just have to call this function and pass a list of images as its inputs. However, you have to consider performing the same image preprocessing steps that were performed before the training procedure. I.e. you have to create Tf-records out of your test images list.
- After creating the tf-records for your test set, convert it to a tensor. And just pass it as a parameter to the function inception_v3 you created in the first step.
- Then you just have to evaluate the bottleneck tensor using a default session.

Actually, the flow of inference from the bottleneck layer I followed is very similar to the evaluation of inception shown in this file: 
https://github.com/tensorflow/models/blob/master/inception/inception/inception_eval.py
But instead of constructing the full graph, I constructed until the bottleneck (avg_pool).

Please let me know if you need more elaboration, and/or code snippets.

**Comment by pronot** [comment]
I still couldn't figure it out. 
I'm trying to retrain model **inception_resnet_v2.py** [https://github.com/tensorflow/models/blob/master/slim/nets/inception_resnet_v2.py](url)
with the retraining code:
[https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/image_retraining/retrain.py](url)
I generated .pb file for inception_resnet_v2.py but i couldn't manage to adapt the retrain.py code for the model. i'm very confused with the code. How can i remove top layer of pretrained model and retrain it with my data?

**Comment by legel** [comment]
@aihamtaleb thanks for those tips. If you do have a code snippet sitting in a file somewhere, would be wonderful to be able to check it out as a gist.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by shlens** [comment]
Due to inactivity, I presume this issue is now resolved but please reopen otherwise.
